modules/python/python-http-header.c

```diff
@@ -0,0 +1,383 @@
+      _py_finish_exception_handling();
+
+      goto exit;
+    }
+
+  PyObject *py_list = (PyObject *) user_data;
+  if (PyList_Append(py_list, py_str) != 0)
+    {
+      gchar buf[256];
+
+      msg_error("Error adding new item to Python List",
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+    }
+
+exit:
+  Py_XDECREF(py_str); // Even if append() succeeds, it doesn't steal the reference to py_str.
+}
+
+static PyObject *
+_py_convert_list_to_pylist(List *list)
+{
+  PyObject *py_list = PyList_New(0);
+  g_assert(py_list);
+
+  if (list)
+    list_foreach(list, _py_append_str_to_pylist, py_list);
+
+  return py_list;
+}
+
+static gboolean
+_py_attach_bindings(PythonHttpHeaderPlugin *self)
+{
+  PyObject *py_args = NULL;
+
+  self->py.class = _py_resolve_qualified_name(self->class);
+  if (!self->py.class)
+    {
+      gchar buf[256];
+
+      msg_error("Error looking up Python class",
+                evt_tag_str("class", self->class),
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+
+      goto exit;
+    }
+
+  py_args = _py_create_arg_dict(self->options);
+  if (!py_args)
+    {
+      gchar buf[256];
+
+      msg_error("Error creating argument dictionary",
+                evt_tag_str("class", self->class),
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+
+      goto exit;
+    }
+
+  self->py.instance = _py_invoke_function(self->py.class, py_args, self->class, self->super.name);
+  if (!self->py.instance)
+    {
+      gchar buf[256];
+
+      msg_error("Error instantiating Python class",
+                evt_tag_str("class", self->class),
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+
+      goto exit;
+    }
+
+  self->py.get_headers = _py_get_attr_or_null(self->py.instance, "get_headers");
+  if (!self->py.get_headers)
+    {
+      msg_error("Error initializing plugin, required method not found",
+                evt_tag_str("class", self->class),
+                evt_tag_str("method", "get_headers"));
+    }
+
+exit:
+  Py_XDECREF(py_args);
+  return self->py.get_headers != NULL;
+}
+
+static void
+_py_detach_bindings(PythonHttpHeaderPlugin *self)
+{
+  Py_CLEAR(self->py.class);
+  Py_CLEAR(self->py.instance);
+  Py_CLEAR(self->py.get_headers);
+}
+
+static void
+_append_str_to_list(gpointer data, gpointer user_data)
+{
+  List *list = (List *) user_data;
+  list_append(list, data);
+}
+
+static void
+_append_headers(PythonHttpHeaderPlugin *self, HttpHeaderRequestSignalData *data)
+{
+  PyObject *py_ret = NULL,
+            *py_list = NULL,
+             *py_args = NULL,
+              *py_ret_list = NULL;
+
+  time_t now = time(NULL);
```

Comment:
The cache needs to be dropped in case of some events. For example after a disconnect reconnect cycle. For example if a remote returns with an autorization error, it would be bad if we cached an invalid credential for an hour.

The reload works nice by the way: after reload, we drop the cache.

Reply:
This is related to keep the http class final question...
Until we think that the http class is a final class then all these 'magics' (signal-slot n mechanism...) are about specializing some parts of the http class... 

So the problem you pointed out is not specific to this PR... 
Of course I can resolve it by adding signals to http response codes (or just error codes) and then cache invalidation can be done in a slot.